SimPEG/ObjectiveFunction.py

```diff
@@ -0,0 +1,258 @@
+                        "Objective Functions must all have the same nP, not "
+                        "{}".format([f.nP for f in objfcts])
+                    )
+                else:
+                    nP = fct.nP
+            self.objfcts.append(fct)
+
+        if multipliers is None:
+            multipliers = len(self.objfcts)*[1]
+        else:
+            for mult in multipliers:
+                assert(type(mult) in self._multiplier_types), (
+                    "Objective Functions can only be multiplied by a float, or"
+                    " a properties.Float, not a {}, {}".format(
+                        type(mult), mult
+                    )
+                )
+            assert len(multipliers) == len(self.objfcts), (
+                "Length of multipliers ({}) must be the same as the length of "
+                "objfcts ({})".format(len(multipliers), len(self.objfcts))
+            )
+        self._multipliers = multipliers
+
+        super(ComboObjectiveFunction, self).__init__(**kwargs)
+        self._nP = nP
+
+    @property
+    def multipliers(self):
+        return self._multipliers
+
+    def _eval(self, x, **kwargs):
+        f = Utils.Zero()
+        for multpliter, objfct in zip(self.multipliers, self.objfcts):
+            f += multpliter * objfct(x, **kwargs)
+        return f
+
+    def deriv(self, x, **kwargs):
+        g = Utils.Zero()
+        for multpliter, objfct in zip(self.multipliers, self.objfcts):
+            g += multpliter * objfct.deriv(x, **kwargs)
+        return g
+
+    def deriv2(self, x, **kwargs):
+        H = Utils.Zero()
+        for multpliter, objfct in zip(self.multipliers, self.objfcts):
+            objfct_H = objfct.deriv2(x, **kwargs)
+            if isinstance(objfct_H, Utils.Zero):
+                pass
+            elif isinstance(objfct_H, Utils.Identity) and self.nP != '*':
+                H += multpliter * sp.eye(self.nP)  # if we need a shape
+            else:
+                H += multpliter * objfct_H
+        return H
+
+    # This assumes all objective functions have a W.
+    # The base class currently does not.
+    @property
+    def W(self):
+        W = []
+        for mult, fct in zip(self.multipliers, self.objfcts):
+            curW = mult * fct.W
+            if not isinstance(curW, Utils.Zero):
+                W.append(curW)
+        return sp.vstack(W)
+
+
+class L2ObjectiveFunction(BaseObjectiveFunction):
+
+    def __init__(self, nP=None, W=None, **kwargs):
+
+        super(L2ObjectiveFunction, self).__init__(nP=nP, **kwargs)
+        if W is not None:
+            if self.nP == '*':
+                self._nP = W.shape[0]
+            else:
+                assert(W.shape[0]) == self.nP, (
+                    'nP must be the same as W.shape[0], not {}'.format(self.nP)
+                )
+        self._W = W
+
+    @property
+    def W(self):
+        if getattr(self, '_W', None) is None:
+            if self._nP != '*' and self._nP is not None:
```

Comment:
right now, nP isn't a `property` so we need to set the private, but we should implement a property for this. 
`Props.Integer` ?